### PR TITLE
New project layout, Carthage, SPM [WIP]

### LINKED ISSUE #79: Create a full test
We should have a repo which is the result of pod-template which specific arguments. Then a test that runs though the same options and then does a diff on output.

Similar to https://github.com/QueryKit/mogenerator-template/blob/master/Makefile#L7-8. We don't need any complex ruby, simple diff does the job.

This should give us more visibility on the [output and how it looks](https://github.com/QueryKit/mogenerator-template/tree/master/Example), and also allows us to know when we unintentionally break something.

### LINKED ISSUE #127: Support for mixed ObjC and Swift Pod?
Point me to a guide if I'm being an idiot. Not much information on this, if we're not creating a pod from total scratch and incorporating someone else's legacy code.

### LINKED ISSUE #131: Recommended project settings
Test case:

1. `pod lib create PROJECT`
2. Open XCode on the workspace
3. Product -> Run

Expected:

Builds with no errors

Actual:

A warning is generated for Pods.xcodeproj titled "Update to recommended settings"

screen shot 2015-11-01 at 4 00 41 pm

### LINKED ISSUE #157: Achieve full compatibility with Swift Package Manager
Update the template layout so that libraries started with `pod-template` will automatically be usable from Swift Package Manager.

TEST CASES:

  * Set up a project with Swift Package Manager like at https://swift.org/package-manager/#example-usage and ensure this library is usable
  * Set up a project with Carthage and ensure no regressions

### PR DESCRIPTION
**Pull request by fulldecent**
# Status

**If you are using Swift 3 and working on a NEW project, check out https://github.com/fulldecent/swift3-module-template it is better than pod-template.**

You can use this _NOW_ for production code if your your project is Swift and has tests. See "How can I play..." below.

PR merge blockers:
- [x] https://github.com/CocoaPods/CocoaPods/pull/5787
- [x] CocoaPods 1.1.0 final -- https://github.com/CocoaPods/pod-template/pull/187#issuecomment-239689605
- [ ] Make it work if "Would you like to do view based testing" is NO (#186)
- [ ] Objective-C support

# How can I play with this PR?
- Clone and `cd` into the directory
- Run some stuff you need to run for any Ruby project:
  - `gem install bundler`
  - `bundle install`
- Run the included test suite:
  - `rake`
  - Note: this uses the latest released version of CocoaPods to run this template (retrieved with Bundler above)
- Try it for real:
  - `cd ~/Desktop`
  - `pod lib create --verbose --template-url='file:///PATH/TO/pod-template' NewPod`
  - Note: this uses your INSTALLED VERSION of `pod` to run this template
- Hacking:
  - **Note: you MUST commit to git BEFORE you do `rake` testing or `pod lib create` testing. Both of those tools pull from master in your local repository. Learn how to squash commits :-)**

---
# PR Improvements to Master
## Wording:
- pod-template -- this project
- template -- a model that can have things changed
- vendored -- a chosen template with variable substitutions and other updates applied and ready for making an iOS framework
## pod-template progress:
- [x] Make vendoring work
- [x] Set up testing with rake/rspec to validate the vendoring process (fixes #79)
- [x] Set up Travis CI
- [x] Update variable substitution process, remove unused code
- [ ] Add testing and make it work if "Would you like to do view based testing" is NO (#186)
## baseline template progress: (new feature)
- [x] Move all files common to Swift and Objective-C into a "baseline" template
- [x] Update to latest gitignore
- [x] Supports Swift Package Manager format (fixes #157)
## swift template progress:
- [x] Create a hello world example (this is a change), it actually does something
- [x] Remove FDChessboardView and my name from project, use variable substitution
- [x] Remove XCode warnings (fixes #131)
- [x] Has "hello world"-level functionality
## objective-c template progress:
- [x] Translate the Swift template into Objective-C (and fixes #127)

### CODE REVIEW

**Comment by fulldecent** [comment]
I have just added objective-c support. A couple notes:

When you create an objective-c pod, the example app is still in Swift. I would argue this is best practice because:

 * Swift is preferred
 * Swift is more restrictive than Objective-C so we are promoting authors to set their Swift compatibility attributes
 * It is easy enough for authors to make their own sample app in Objective-C if they really want to
 * It makes this repository simpler

For the same reasons, we also have Swift tests when you have an Objective-C pod.

Overall, the message is: if you are making an Objective-C pod in 2016 you should be thinking about all the other Swift users that will be using your pod.

**Comment by orta** [comment]
Agreed re: the Swift vs Obj-C stuff

**Comment by orta** [comment]
I'm struggling to test this with the `file://` way, I keep getting 

``` sh
2016-08-05 19:19:16.354 ruby[17944:6516871] CFPropertyListCreateFromXMLData(): Old-style plist parser: missing semicolon in dictionary on line 204. Parsing will be abandoned. Break on _CFPropertyListMissingSemicolon to debug.
/Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/plist/ffi/core_foundation.rb:271:in `ptr': uninitialized constant CoreFoundation (NameError)
	from /Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/plist/ffi/core_foundation.rb:73:in `RubyHashPropertyListRead'
	from /Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/plist/ffi.rb:54:in `read_from_path'
	from /Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/plist.rb:22:in `read_from_path'
	from /Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/project.rb:200:in `initialize_from_file'
	from /Users/orta/.gem/ruby/2.1.3/gems/xcodeproj-1.2.0/lib/xcodeproj/project.rb:102:in `open'
	from /Users/orta/Desktop/ok_then/configurator/ProjectManipulator.rb:30:in `run'
```

Which is odd, because if I manually then try open the file in pry it works just fine. Will have to try on another computer to see if I get the same issue.

File: templates/objective-c/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+```bash
+$ gem install cocoapods
+```
+
+To integrate PROJECT into your Xcode project using CocoaPods, specify it in your `Podfile`:
+
+```ruby
+source 'https://github.com/CocoaPods/Specs.git'
+platform :ios, '9.0'
+use_frameworks!
```
Comment (orta) [inline templates/objective-c/README.md]: these two are optional, and are mostly not required - should probably be removed

```
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'
```

File: templates/objective-c/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# PROJECT
+
+[![CI Status](http://img.shields.io/travis/fulldecent/PROJECT.svg?style=flat)](https://travis-ci.org/fulldecent/PROJECT)
+[![Version](https://img.shields.io/cocoapods/v/PROJECT.svg?style=flat)](http://cocoadocs.org/docsets/PROJECT)
+[![License](https://img.shields.io/cocoapods/l/PROJECT.svg?style=flat)](http://cocoadocs.org/docsets/PROJECT)
+[![Platform](https://img.shields.io/cocoapods/p/PROJECT.svg?style=flat)](http://cocoadocs.org/docsets/PROJECT)
```
Comment (orta) [inline templates/objective-c/README.md]: would recommend these URLs go to CocoaPods.org - probably a better URL for this stuff: `http://cocoapods.org/pods/PROJECT`

File: templates/objective-c/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# PROJECT
+
+[![CI Status](http://img.shields.io/travis/fulldecent/PROJECT.svg?style=flat)](https://travis-ci.org/fulldecent/PROJECT)
```
Comment (orta) [inline templates/objective-c/README.md]: I wonder if we can do something about the `fulldecent` here, and below in the Carthage section

File: templates/objective-c/Source/Info.plist [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (orta) [inline templates/objective-c/Source/Info.plist]: This image is cute ^

**Comment by dbgrandi** [comment]
I just tried and got the same error that @orta mentioned [above](https://github.com/CocoaPods/pod-template/pull/187#issuecomment-237989852)

File: templates/objective-c/README.md (hunk 1 above)
Comment (fulldecent) [inline templates/objective-c/README.md]: Thank you, fixed in ad39d3612fc60dfa3a9a66d8b684b852fca67c48

File: templates/objective-c/README.md (hunk 2 above)
Comment (fulldecent) [inline templates/objective-c/README.md]: Fixed in 68830d8920b4b7fc9b6b2db3afa8f773ec90ad1c

File: templates/objective-c/README.md (hunk 3 above)
Comment (fulldecent) [inline templates/objective-c/README.md]: Fixed in 8657fe8b77f501dfc747943ad3331733ee8a3c07

**Comment by fulldecent** [comment]
@orta RE https://github.com/CocoaPods/pod-template/pull/187#issuecomment-237989852 @dbgrandi RE https://github.com/CocoaPods/pod-template/pull/187#issuecomment-237996910

Please see the successfully running build at https://travis-ci.org/CocoaPods/pod-template/jobs/141253669

Newer builds are failing due to a change in Travis stack versions, I am still unsure what is going on with that.

**Comment by fulldecent** [comment]
Is anyone else able to run / not able to run this branch?

**Comment by azeff** [comment]
Just tried and got same CFPropertyListCreateFromXMLData error.

**Comment by fulldecent** [comment]
@azeff @dbgrandi @orta Thank you, the latest two commits fixes that problem for me when using a fresh dev box. Could you please confirm?

**Comment by orta** [comment]
Yep - I'm getting further! I'm now getting:

```
/Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:85:in `gsub!': invalid byte sequence in UTF-8 (ArgumentError)
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:85:in `block in replace_variables_in_files'
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:82:in `glob'
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:82:in `replace_variables_in_files'
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:60:in `block in run'
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:59:in `chdir'
	from /Users/orta/dev/ios/NewProjectOKTMP/configurator/MainConfigurator.rb:59:in `run'
	from ./configure:9:in `<main>'
```

Which [I think](https://robots.thoughtbot.com/fight-back-utf-8-invalid-byte-sequences) can be fixed by doing 

```ruby
    def replace_variables_in_files
      Dir.glob("**/*") do |file_name|
        next if File.directory?(file_name)
        text = File.read(file_name).encode('UTF-8', 'binary', invalid: :replace, undef: :replace, replace: '')
        [...]

```

**Comment by fulldecent** [comment]
@orta Thank you, added

**Comment by orta** [comment]
Runs through fine 👍 will put some time in tomorrow testing

**Comment by fulldecent** [comment]
Excellent, thank you! Build is passing again which makes me happy.

**Comment by orta** [comment]
@segiddins / @benasher44 - do you think CP 1.1.0 will include linking Pods to frameworks? 

e.g. if I have this?

screen shot 2016-08-14 at 2 13 07 pm

Could I like `Quick`?

Otherwise, running this without a demo would mean that you get no Pods set up at all. 
Which I may be able to live with, but you'll see two places for writing tests: inside the framework, and inside the example app. One of which is set up as you asked, the other which isn't.

**Comment by orta** [comment]
Pushed a few minor commits ( notably one that stops Pod names with spaces ) for some minor style things while digging aroubnd

**Comment by benasher44** [comment]
@orta 1.1.0.beta.1 will work with that setup if you add the `ThereTests` to your Podfile. If you don't add `ThereTests` to your Podfile, and you use CocoaPods master (which will be 1.1.0.beta.2?), you'll get this warning (from CocoaPods/CocoaPods#5733) but the project should still build:

```
[!] The Podfile contains framework targets, for which the Podfile does not contain host targets (targets which embed the framework).
If this project is for doing framework development, you can ignore this message. Otherwise, add a target to the Podfile that embeds these frameworks to make this message go away (e.g. a test target).
```
In short, this should work in 1.1.0, but it doesn't work in the current beta unless you add `ThereTests` to your Podfile.

**Comment by orta** [comment]
Interesting, would you take a PR adding an ENV var check before showing that message? In this case you can be guaranteed to be wanting that.

/cc @segiddins

**Comment by benasher44** [comment]
Does an ENV var mean doing something like this: `INSTALL_FOR_FRAMEWORK=1 pod install`? If so, then the intent of the Podfile you're installing with there is to always integrate pods with a framework/library, so maybe this should just be Podfile DSL?

**Comment by azeff** [comment]
Failed to create lib without demo project.
```
PATH/configurator/ProjectManipulator.rb:59:in `remove_demo_project': undefined method `remove_from_project' for nil:NilClass (NoMethodError)
	from PATH/configurator/ProjectManipulator.rb:31:in `run'
	from PATH/configurator/SwiftConfigurator.rb:54:in `perform'
	from PATH/configurator/SwiftConfigurator.rb:13:in `perform'
	from PATH/configurator/MainConfigurator.rb:55:in `run'
	from ./configure:9:in `<main>'
```

Tried to build create project and failed with following.
```
While reading <PATH>/Resources/wk.png pngcrush caught libpng error:
   Not a PNG file..
```
Think problem is that `MainConfigurator.replace_variables_in_files` runs for binary files as well.

**Comment by fulldecent** [comment]
@azeff `replace_variables_in_files` with binary may be an issue. But that code runs after the crash from `SwiftConfigurator.perform(configurator: self)`.

For this crash, I am hoping to get some help from the Xcodeproj team. Maybe @fabiopelosin or @alloy could help :-)?

**Comment by azeff** [comment]
@fulldecent sorry for misleading, issue with PNG is not related to a crush and I see it on a lib created with default parameters (by pressing enter each time it asks). Try to build for device or open `wk.png` from the project.

**Comment by orta** [comment]
Just to make sure it doesn't look like I abandoned this, it's now blocked on a release of CP 1.1.0, in order for this implementation of pod-template to work we need the ability to add pods to a framework. 

I think we'll also need to make a change to CocoaPods to pass along the version so that we can let people know that it's only supported when working against 1.1.0

**Comment by fulldecent** [comment]
@orta I have added a blocker for https://github.com/CocoaPods/CocoaPods/pull/5787 in the notes above. CocoaPods 2.1.0 doesn't even have a milestone yet. Do you have a specific issue we can reference as a blocker?

**Comment by orta** [comment]
This is the issue, see the discussion above with ben + I, https://github.com/CocoaPods/pod-template/pull/187#issuecomment-239689605 ( It's for CP 1.1.0, which is currently in beta )

**Comment by fulldecent** [comment]
Thank you, blocker added

**Comment by fulldecent** [comment]
FYI, this PR is usable for production code today -- assuming you use Swift and testing. I have updated notes above for same.

**Comment by fulldecent** [comment]
Most people that are making new pods will probably want to use https://github.com/fulldecent/swift3-module-template

Long term, I would like to replace CocoaPods/pod-template with what is at swift3-module-template and cover all the other use cases that CocoaPods/pod-template does.

**Comment by orta** [comment]
1.1.0 is now out, and we're good to start thinking about getting this in

**Comment by orta** [comment]
I've my plan is to make a CP Mac release at the weekend, and get this out, do you want to consider writing a blog post for this? I'd love to get your name out about the changes

**Comment by fulldecent** [comment]
# Update to CocoaPods Pod Template 

*by William Entriken, posted 2016-10-21*

The newest CocoaPods release 1.1.1 is now compatible with the updated Pod Template and it brings new features to all pod authors!

Pod Template (`pod-template`) is used when you run `pod lib create MyNewAwesomePod`. It is an opinionated starting point for all pod authors and it gives you everything you need to start delivering a nice and reusable pod.

All the best practices for pod authors are already built in. But the 1.1.1 release bring many new features. To recount, here is everything:

 * Prompt-based project generation
 * Support for CocoaPods, Carthage and project layout compatible with Swift Package Manager
 * Clean folder structure
 * MIT license
 * Testing as a standard
 * Turnkey access to Travis CI

As new versions of Xcode are released and as Swift evolves, Pod Template must be updated to take advantage of new features. You will get all the latest updates every time you run `pod lib create` command. For more information and to help shape the future of modular development in Swift & Objective-C see the project page at https://github.com/CocoaPods/pod-template

screen shot 2016-10-20 at 7 11 52 pm

**Comment by fulldecent** [comment]
@orta Thank you! Here is a blog post.

**Comment by orta** [comment]
Yep, looks like that PNG gets corrupted, probably is getting find and replaced on it's internals. 

Looks like all this swift tuff needs updating for Xcode 8, sigh

**Comment by orta** [comment]
Hrm, png still corrupting but at least project removal seems to work fine now and Swift seems to work fine on v3

**Comment by orta** [comment]
Looks like the Xcode project for Objective-C libraries doesn't open, and is somehow correupted

**Comment by fulldecent** [comment]
https://github.com/fulldecent/swift3-module-template is now compatible with CocoaPods. It can be run directly with

    pod lib create --verbose --template-url='https://github.com/fulldecent/swift3-module-template.git' NewPod

For new Swift 3 projects, this is superior. It is very clean, has no dependencies and is reproducibly documented.

Please help review. If there is interest, I will make a PR here.

**Comment by orta** [comment]
Just a reminder for anyone interested in this PR, this is still blocked from a merge on Objective-C support, and work on the guides to reflect new changes.

**Comment by jugutier** [comment]
@fulldecent Is this up to date with your repository? I'll look at the merge conflict with the ObjC branch since I worked on it last.

**Comment by jugutier** [comment]
On the left is the result of:

```pod lib create --template-url='file:///path/to/this-repo-develop/pod-template' MyPod```

for Objective-C, Demo=Yes,Test=Quick,ViewBased=No

Right is the same but with swift.

Few things I noticed:

1. Obj-C is broken which is not news.
2. The App target/project now goes after the development pod target and is also called 'iOS Example'. Any reasons to change it from the way it was before? (My point is why introduce breaking changes into a project structure that has been used and that looks more to the structure of the user of any given pod)
3. MyPod.xcodeproj is also referenced within 'iOS Example' this project reference isn't currently done on master. Any reasons for this?
4. `baseline/.travis.yml` has a flag for `-enableCodeCoverage YES` which I think is to run Swift 2.3 tests on Swift 3 compiler? if that's the case and it's hardcoded it might not work for Obj-C. Coments pls 🔢 

screen shot 2017-02-07 at 11 13 36 pm

**Comment by fulldecent** [comment]
@jugutier This PR is live and still valid. If we can get it working with ObjC then we should probably merge it because the new configuration engine is nice. This PR still suffers from the main problem of `pod-template`, which is that it needs to be painfully updated for each Xcode release. Therefore I have not been involved in this PR.

The swift3-module-template has a dead-simple configurator, no external dependencies and a *recipe*. This means generating the template is repeatable. That solves the problem of updating for each Xcode release.

I have just now created https://github.com/fulldecent/objective-c-module-template as a starting point to do the same thing for Objective-C projects. If someone can help with this new part then we will have a complete replacement for pod-template.

**Comment by jugutier** [comment]
@fulldecent I am already doing it here: https://github.com/jugutier/pod-template/tree/develop

my plan is to rebase master from this branch to get the latest that happened there (mostly my changes for the obj-c template) but for that I'd like to understand points 2, 3 and 4 above. Could you see if you have some input?

Thanks!

**Comment by fulldecent** [comment]
Hello all, I have been working on this PR for almost a year. Thank you for your patience, it's worth the wait!

After lots of testing, dogfooding and feedback, please see https://github.com/CocoaPods/pod-template/pull/210

I put all my heart into that PR and it supersedes this #187 so I'm closing this one. Specifically, the new one now supports Swift AND Objective-C. It was such a simple fix... just make a mixed project. But now it works and so it is a realistic contender to replace the old PR.

Would you please bring your comments from this thread and consider your same issues on the new PR. As always, I'm very eager for your feedback.

**Pull request closed without merging**